src/doc/src/reference/features.md

```diff
@@ -1,134 +1,464 @@
+  viewing the entire graph can be quite large and overwhelming. Use this when
+  you are trying to figure out which features are enabled on a specific
+  package and why. See the example at the bottom of the [`cargo tree`] page on
+  how to read this.
+
+[`cargo tree`]: ../commands/cargo-tree.md
+
+### Feature resolver version 2
+
+Cargo has a new feature resolver which uses a different algorithm for
+resolving which features are enabled. This can be enabled by specifying the
+resolver version in `Cargo.toml` like this:
+
+```toml
+[package]
+name = "my-package"
+version = "1.0.0"
+resolver = "2"
+```
+
+This new resolver avoids unifying features in a few situations where that
+unification can be unwanted. The exact situations are described in the
+[resolver chapter], but in short, it avoids unifying:
+
+* Target-specific dependencies for targets not currently being built.
+* [Build-dependencies] and proc-macros do not share features with normal
+  dependencies.
+* [Dev-dependencies] do not activate features unless building a target that
+  needs them (like tests or examples).
+
+Avoiding the unification is necessary for some situations. For example, if a
+build-dependency enables a `std` feature, and the same dependency is used as a
+normal dependency for a `no_std` environment, enabling `std` would break the
+build.
+
+However, one drawback is that this can increase build times because the
+dependency is built multiple times (each with different features). When using
+the new resolver, it is recommended to check for dependencies that are built
+multiple times to reduce overall build time. If it is not *required* to build
+them with separate features, consider adding features to the `features` list
+in the [dependency declaration](#dependency-features) so that the duplicates
+end up with the same features (and thus Cargo will build it only once). You
+can detect these duplicate dependencies with the [`cargo tree
+--duplicates`][`cargo tree`] command. It will show which packages are built
+multiple times; look for any entries listed with the same version. See
+[Inspecting resolved features](#inspecting-resolved-features) for more on
+fetching information on the resolved features. For build dependencies, this
+is not necessary if you are cross-compiling with the `--target` flag because
+build dependencies are always built separately from normal dependencies in
+that scenario.
+
+The `resolver = "2"` setting also changes the behavior of the `--features` and
+`--no-default-features` [command-line options](#command-line-feature-options).
+Previously, you could only enable features for the package in the current
+working directory, regardless of which `-p` flags were used. With `resolver =
+"2"`, the flags will enable the given features for the packages selected on
+the command-line with `-p` flags. For example:
+
+```sh
+# This command is now allowed, regardless of which directory you are in.
```

Comment:
Using "now" sets the frame of temporal reference and will look odd when there is v.2.x or some later version. Would it be better to use direct version references, e.g. "This command is allowed in resolver 2" / "not allowed in resolver 1"? There are a few other places in the text where "now", "previously" and "new" are used in a similar way. This is really nit-picking on my side, sorry.